### ISSUE TITLE
Kopf adding removing finalizerr on some custom resources

### ISSUE THREAD

**Issue by jamiechapmanbrn** [nolar/kopf#587]
We recently upgraded from kopf 0.25 to 0.28. For whatever reason it's modifying some custom resources constantly, adding and removing the finalizer. This is hitting our customresource cockroachdb, seemingly exclusively.

When this happens it's sending an update through to eventhandler 

cockroachdb's spec looks like this:
```
---
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: cockroachdbs.agilicus.com
spec:
  scope: Namespaced
  group: agilicus.com
  versions:
    - name: v1
      served: true
      storage: true
  names:
    kind: cockroachdb
    plural: cockroachdbs
    singular: cockroachdb
  additionalPrinterColumns:
    - name: DBName
      type: string
      priority: 0
      JSONPath: .spec.dbname
      description: The name of the cockroachdb database
    - name: User
      type: string
      priority: 0
      JSONPath: .spec.user
      description: The initial user of the cockroachdb database
    - name: Restore
      type: boolean
      priority: 0
      JSONPath: .spec.restore
      description: Enable Restore from database
    - name: RestoreFrom
      type: string
      priority: 0
      JSONPath: .spec.restorefrom
      description: Restore database from backup based on date/time (dateparser)
```

The watcher looks like this:
```
        @kopf.on.event(
            "agilicus.com",
            "v1",
            "cockroachdbs",
            registry=self.registry,
        )
        def event(logger, event, body, status, **_):
            self.send_event(
                .
                .
                .
            )
```

* Kopf version: 0.28
* Kubernetes version: v1.18.9
* Python version: 3.8
* OS/platform: ubuntu

**Comment by jamiechapmanbrn** [comment]
The issue is that we are watching with two kopfs here:
The cockroachdb operator is written w/ kopf and should be adding the finalizer and handling the object
This operator which was just upgraded to 0.28 watches and creates cockroachdbs as a larger loop to create agilicus customer resources, like for example a database.

For now we can roll kopf in the outer operator to prevent conflict, but there should be a simple way to work around this.

**Comment by nolar** [comment]
Thanks for reporting this case. Can you please clarify several moments?

Do I get it right, that you have 2 kopf-based operators? One is shown in the code snippet with `on.event()`, and another does some more advanced things — is this correct? 

Is it possible for you to share the decorators of that other operator? I.e. `@on.create/update/delete/daemon/timer/event` lines — without the function names or content or anything significant?

---

The constant removal/addition of finalizers (and, in some cases, last-seen annotations) are known to happen when 2+ operators fight for the same resource while being unaware of each other. A typical solution would be to give them different identities — i.e. annotation prefixes/names and different finalizer names.

The identities can be changed via `@on.startup` handlers by editing `settings.persistence.finalizer` (and also progress- & diffbase-storages, but they are less relevant here). See this section:

* https://kopf.readthedocs.io/en/stable/configuration/#finalizers

I can even presume that it is safe to change the finalizer on the existing operator. It will cause some initial activity (which you probably already have) for placing/removing proper finalizers. But do not trust me here — it is better if you test this in some safe environment.

**Be careful** with the diffbase storage — changing it without backward compatibility in mind would re-process all your resources as they have been just created. The smooth upgrade scenario via a 2-storage temporary setup is also documented:

* https://kopf.readthedocs.io/en/stable/configuration/#storage-transition

---

I can now blindly assume that _maybe_ the on-event-containing operator decides that there should be no finalizer, sees one there, and removes it. The other operator knows that there should be a finalizer (same-named), sees that it is gone, and re-adds it. And so on forever. To verify that, I need some overall information on that "other" operator requested above. 

Nevertheless, this is the desired behaviour — mostly for the cases when some resources change the state so that they stop matching the filters of the finalizer-relevant handlers (specifically, on-deletion, daemons, timers). For example: draft an operator with only one handler: `@on.delete(..., labels={'x': 'y'})` — and then manually add/change/remove that label with that value — the finalizer will be added/removed accordingly.

With all the complexity that the handlers' filters introduce, I am currently not sure if there is a way to decide in advance that the operator is always and surely finalizer-free, so that it could afford to ignore the finalizer completely.

It can be much easier solved by configuring the finalizer names instead. This is actually the "proper" way for multiple overlapping operators — to have their own identities.

**Comment by jamiechapmanbrn** [comment]
Yup, you've got it correct.

The cockroach operator uses these decorators:
@kopf.on.create("agilicus.com", "v1", "cockroachdbs")

@kopf.on.delete("agilicus.com", "v1", "cockroachdbs")

That on.delete decorator I'm sure is what's fighting with the on.event decorator used in the other operator.

For describing our use case a little more here
I'll call the first operator mentioned the 'outer operator', for which I've pasted the code in the first comment
The code I've pasted here is for the 'inner operator'.

The inner operator is the one that actually does the work for this custom resource, it create cockroachdb databases for us when requested, and will log when they are deleted.

The outer operator, is merely using this resource to do other work. It calls another function to get all the resources it expects an environment to have given some crud resources. For example it might include a deployment, a secret and a cockroachdb. The outer operator is really only watching what resources exist so that it knows what it needs to create, destroy and modify. We really don't want it to try to 'own' any of these resources.

It sounds like changing the finalizer name in the outer operator  is the solution to go with here. Thanks for your help!
